### ISSUE TITLE
Higlighting issue with strings & comments

### ISSUE THREAD

**Issue by apfelbox** [PrismJS/prism#307]
This is the combined issue for all related issues (#187, #183, #155, #153).

There is a fight between the highlighting of strings vs. comments.
Currently only one of them can work

```
// this is "a string"
"string with /* comment";
```

I will take a look whether this issue can be "fixed" (it can't be fixed for all cases due to the limited power of regular expressions though).

**Comment by foo123** [comment]
Check [prism-grammar](https://github.com/foo123/prism-grammar) external parsing add-on for Prism, which uses a grammar to highlight solving all these issues (and more)

**Comment by WebReflection** [comment]
chiming in with a proof of concept that I've no idea if is worth exploring.

Basically, it removes strings, highlight them, then put them back.

```js
function safeHighlight(content, highgligh) {
  var str = [];
  function pre(content) {
    return content.replace(
      /(["'])(?:(?=(\\?))\2.)*?\1/g,
      ($0, $1) => $1 + (str.push($0.slice(1, -1)) - 1) + $1
    );
  }
  function post(content) {
    return content.replace(
      /(["'])(\d+)\1/g,
      ($0, $1, $2) => $1 + str[$2] + $1
    );
  }
  return post(highgligh(pre(content)));
}

var content = 'var str = "hello/**/world//!";';

// use prism or the actual highlighter instead of String
safeHighlight(content, String);
```
I haven't done many benchmarks but it seems like a simple/safe solution for strings in comments and/or in code: it will put them back in their context, if that was highlighted, strings will result highlighted too.

Best Regards

**Comment by zeitgeist87** [comment]
@WebReflection your solution would certainly work for strings, although it seems more like a hack. Do you have any proposal for turning it into a general purpose solution?

We have recently merged a partial solution for that problem #779. It can correctly handle strings with one comment inside but not more (e.g.: `"hello/**/world!"`)

**Comment by WebReflection** [comment]
well, the general purpose solution is what you call the hack:

  * find all strings and put a string placeholder
  * parse everything without problems and ignoring all strings content
  * put strings back using the placeholder

A char-by-char parser wouldn't have this problem because it would know it's inside a string ... my work-around does exactly the same making the "_inside a string_" a no-brainer for the parser.

I could write a more robust example with a better placeholder but I honestly don't know how to improve every language parsed by `prism`

**Comment by zeitgeist87** [comment]
> well, the general purpose solution is what you call the hack:
> 
> find all strings and put a string placeholder
> parse everything without problems and ignoring all strings content
> put strings content back using the placeholder

I understand how it works and it is a nice solution. With general purpose solution I mean something that works for any grammar and not only for strings but also for other stuff. This problem is not limited to strings. For example you can have a regex with strings inside `/something"foo"something"bar"/`

**Comment by WebReflection** [comment]
> For example you can have a regex with strings inside /something"foo"something"bar"/

not an issue, regexp will be highlighted as such, and strings will be put back.
So this: `/something"foo"something"bar"/` will become `/something"_@no\x01shenanigans@_"something"_@no\x01shenanigans@_"/` and then it will become again the original string.

TL;DR it should work universally but the order obviously matters.

  * `RegExp` don't suffer the code bug because both multiline and single line needs to be escaped, so comments first
  * regular expressions would be second and the same trick could be done for literal `RegExp` too ... you remove strings first, then literal RegExp, then you parse
  * string parsing at the end because it's not safer, easier

This would be my order/solution ... but yet it's crafted for JS comments and JS strings.

**Comment by WebReflection** [comment]
to better explain:

  * pre process dropping strings
  * process processed outcome finding comments, literal regexp, rest of the code, strings
  * put all strings back in place so those within comments will be comments, those within regexp will be regexp, those within strings will be strings

If needed, drop strings entirely (remove $1 too and return just a placeholder) and put them back at the end using a `restore` callback that put back them highlighted.

Does this make sense?

**Comment by foo123** [comment]
Since i happen to get notified in this discussion, let me enter my 2 cents

Yes using placeholders partialy solves the problem, so in this sense it is a hack, a nice hack though.

For example multiply escaped strings, inside strings e.g `"a striung with \\\\\\"escaped strimng\\\\\\\\\\" inside"` (used in javascript) will fail, and so on.. A major problem is escaping, which needs char-by-char parsing (at least in such cases) since regexps cannot count parentheses, escapes and so on..

Then other combinations, e.g strings inside comments, or vice-versa, or comments inside regexps and so on.. Or tags inside CDATA and so on..

So this is not a good solution, only partial and even this in some cases.

i still suggest using a special (delimited) tokenizer which handles only delimited tokens (e.g comments, strings, regexps, cdata and what not, which are all over the place). it is not difficult, it is not slower (in fact it will be faster) and is actually intuitive to define in a specification e.g

```javascript
"comment" : ["/*","*/"],
"string": ["'","'","\\"],
"regexp": ["/","/","\\"]
// and so on..
```
Cheers

**Comment by zeitgeist87** [comment]
> i still suggest using a special (delimited) tokenizer which handles only delimited tokens (e.g comments, strings, regexps, cdata and what not, which are all over the place). it is not difficult, it is not slower (in fact it will be faster) and is actually intuitive to define in a specification e.g

I've already implemented exactly that some time ago, but the code size got too big. The problem is that `prism-core.js` has to stay small and light-weight, because that is one of the major features of Prism.

**Comment by zeitgeist87** [comment]
> Does this make sense?

Well it seems to get quite complicated. The constraints for performance and code size are quite challenging here. We already had a few proposed solutions that didn't cut the mustard. Nevertheless I am happy to test and review any proof of concept implementations...

**Comment by WebReflection** [comment]
@foo123 your example is broken, this string doesn't create a `SyntaxError`: `"a string with \\\\\\\"escaped string\\\\\\\\\\\" inside"`

Moreover, that's not an issue at all with the `RegExp` I've used.

```js
var code = 'var s = "a string with \\\\\\\"escaped string\\\\\\\\\\\" inside";';
code.replace(/(["'])(?:(?=(\\?))\2.)*?\1/g, '$1hello$1');
// "var s = "hello";"
```

So I'm not sure what are your concerns about correctly escaped strings.

For non correctly escaped strings though, well ... that's a "_who cares_" case because the code is broken and cannot be possibly highlighted correctly, right?

**Comment by foo123** [comment]
The fact remains that regexps cannot count (e.g number of parentheses, or number of escapes and so on..) So if they cannot count, all delimited tokens that can be escaped , meaning they can contain their own end-pattern if it is escaped in a perfectly valid way, cannot be parsed correctly with regexps. Maybe i phrased it better now.

**Comment by WebReflection** [comment]
I agree char-by-char is a rock-solid solution but still I'd like to underline the `RegExp` I'm using works with escaped strings, no matter how many escapes, because the `RegExp` I'm using **does count** escapes.

It's not a universal solution, I agree, but it works with strings and with literal `RegExp`.

Here an improved example that make the intermediate content easily parsable via `prism`

```js
var safeHighlight = (function () {
  var
    uid = '_@no\x01shenanigans@_',
    findRe = /(\/)[^/*](?:(?=(\\?))\2.)+\1/g,
    findStr = /(["'])(?:(?=(\\?))\2.)*?\1/g,
    restoreRe = new RegExp('(\\/)' + uid + '(\\d+)\\1', 'g'),
    restoreStr = new RegExp('(["\'])' + uid + '(\\d+)\\1', 'g'),
    dropRe = function ($0, $1) {
      return $1 + uid + (re.push($0.slice(1, -1)) - 1) + $1;
    },
    dropStr = function ($0, $1) {
      return $1 + uid + (str.push($0.slice(1, -1)) - 1) + $1;
    },
    placeRe = function ($0, $1, $2) {
      return $1 + re[$2] + $1;
    },
    placeStr = function ($0, $1, $2) {
      return $1 + str[$2] + $1;
    },
    str, re
  ;
  return function safeHighlight(content, highgligh) {
    str = [];
    re = [];
    var out = highgligh(
      content
        .replace(findStr, dropStr)
        .replace(findRe, dropRe)
    )
      .replace(restoreRe, placeRe)
      .replace(restoreStr, placeStr)
    ;
    str = [];
    re = [];
    return out;
  };
}());

var code = `
  /* this is a multiline comment */
  var s = "a string with \\\\\\\"escaped string\\\\\\\\\\\" inside";
  var c = "a string with /*ml comment*/ and a //sl too";
  // this is a single line comment
  var re = /this is a re" with a string "inside/g;
  var rec = /this is a re with a [^/*ml*/] inside/g;
`;

safeHighlight(code, String) === code; // true
```

How does the intermediate code look like?
```js
/* this is a multiline comment */
var s = "_@noshenanigans@_0";
var c = "_@noshenanigans@_1";
// this is a single line comment
var re = /_@noshenanigans@_0/g;
var rec = /_@noshenanigans@_1/g;
```

Is that middle-content easy to parse for `prism`? I think so.

Maybe I should  put a jsbin online so we can all judge and find possible problems?

**Comment by foo123** [comment]
Yes having the regexp take account of any preceding escapes does fix the thing, however i;m not sure what would happen with multiple escaped strings inside a single string, does it work the same? It should be tested. true not usual case, but lets do it for science :)

Btw even ace editor (if i'm not mistaken) which github uses to highlight code has a hard time with these things even though it uses a state machine for parsing tokens (ace javascript mode)

**Comment by WebReflection** [comment]
dunno, what I think is good, is that I've created a [jsbin](https://output.jsbin.com/gifagiqaru) for this case.

At least, if nothing, there is a playground to test nasty code and see how to proceed.

There are surely edge cases, but since edge cases are a "_won't fix_" here, due nature of the parser which is `RegExp` based and, as such, cannot be perfect, I guess at least reducing them would be already a good achievement.

Hopefully worth exploring.

**Comment by WebReflection** [comment]
actually forgot to escape html to avoid gotchas when strings/regexps are back

**Comment by zeitgeist87** [comment]
@WebReflection Your expanded solution works as a proof of concept, but you have to fully integrate it into `prism-core.js`, which I believe is possible and could work. However I am still skeptical if it would be any better than the solution we already have. The input for your function must be an arbitrary Prism grammar for an arbitrary language. It shouldn't be limited to JavaScript.

Also the string inside of a regex example `var re = /this is a re" with a string "inside/g;` doesn't work, because `placeRe` encodes the `"` characters to `&quot;`. Therefore they cannot be found any more by the `restoreStr` regex.

> There are surely edge cases, but since edge cases are a "won't fix" here, due nature of the parser which is RegExp based and, as such, cannot be perfect, I guess at least reducing them would be already a good achievement.

We have recently merged a general purpose, small solution with minimal performance impact, which reduces these edge cases significantly #779.

**Comment by WebReflection** [comment]
nice catch about the regExp, it's because I've adde the HTML escaper without retesting (stupid me).

I understand it should work as general purpose solution but I don't think it will be accepted here (because of skepticism) so I won't bother with a PR but I'll fix my proof of concept.

If there is real interest in having it in I'll try to improve the proof of concept.

**Comment by WebReflection** [comment]
FYI I've fixed the issue with strings in regexps (with strings in general) https://jsbin.com/qazida / @zeitgeist87

**Comment by LeaVerou** [comment]
The problem with removing strings and using placeholders is that it doesn't handle all other kinds of nested tokens, and if we handle those too we're basically parsing the input, so we may as well do some proper lightweight parsing first.
I wonder how much that would slow us down, actually. We could have a special kind of token that is nested (has a start and an end delimiter), then go over the input once and figure out what the groups are. Done right it might be a workable solution, perhaps.

**Comment by WebReflection** [comment]
I'm not sure you should handle any nested token within single or double quoted strings but I've improved the example with better code, Python and Ruby compatible too (easy to make it safer with others) and improved a bit in terms of API
https://github.com/PrismJS/prism/issues/307#issuecomment-205453544

The previous version was suffering nested tokens indeed, now in the worst case scenario you gonna have a number highlight instead of string.

Still not perfect but we all know with RegExp only it's hard to be so :-)

**Comment by Golmote** [comment]
This is mostly fixed by the greedy feature (#779) and its improvement (#967). I'm closing the issue.

**Comment by ziagit** [comment]
Hi there,
I'm using Prism with Laravel, 
The problem I face; when I have something like {{myText}} in my HTML file, this string interpolation cause, not to show the whole page,  

please help
thanks
